plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUClass.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.PsiClass
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiIdentifier
+import org.jetbrains.kotlin.asJava.classes.KtLightClass
+import org.jetbrains.kotlin.fir.declarations.FirClass
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getOrBuildFirOfType
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getResolveState
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.uast.*
+import org.jetbrains.uast.internal.acceptList
+import org.jetbrains.uast.visitor.UastVisitor
+
+sealed class AbstractFirKotlinUClass(
+    givenParent: UElement?
+) : FirKotlinAbstractUElement(givenParent), UClass, UAnchorOwner {
+    override val uAnnotations: List<UAnnotation>
+        get() {
+            // TODO: Not yet implemented
+            return emptyList()
+        }
+
+    override val uastDeclarations: List<UDeclaration> by lz {
+        mutableListOf<UDeclaration>().apply {
+            addAll(fields)
+            addAll(initializers)
+            addAll(methods)
+            addAll(innerClasses)
+        }
+    }
+
+    open val ktClass: KtClassOrObject?
+        get() = (psi as? KtLightClass)?.kotlinOrigin
+
+    // TODO: it's not yet clear whether this is needed
```

Comment:
I expect part of inner-class, superclass, fields, methods retrieval can benefit from this, but I'm not quite sure yet.

Reply:
It depends on how for instance properties are represented in FIR? If as a pair of getter and setter - then definitely Uast will benefit

Reply:
It is a set of backing field (if any), getter and setter, so I think we'll benefit from FirProperty (or KtProperty) retrieval from FirRegularClass (or KtClassLikeSymbol).